### PR TITLE
Make it easy to ignore direct subclasses of a class in dead code plugins

### PR DESCRIPTION
**Pull request by Morriar**
This PR introduces the `ignore_classes_inheriting_from`  DSL:

```rb
module Spoom
  module Deadcode
    module Plugins
      class MyPlugin < Base
        ignore_classes_inheriting_from("Foo", /Bar.*/)
      end
    end
  end
end
```

Will ignore classes such as `class X1 < Foo` or `class X2 < Bar42`.

### CODE REVIEW

File: lib/spoom/deadcode/plugins/base.rb [hunk 1]
```diff
@@ -35,6 +35,24 @@ def ignore_class_names(*names)
...
+          # class MyPlugin < Spoom::Deadcode::Plugins::Base
+          #   ignore_subclasses_of(
+          #     "Foo",
+          #     "Bar",
+          #     /Baz.*/,
+          #   )
+          # end
+          # ~~~
+          sig { params(names: T.any(String, Regexp)).void }
+          def ignore_subclasses_of(*names)
```
Comment (egiurleo) [inline lib/spoom/deadcode/plugins/base.rb:52]: Nit -- do you think `ignore_direct_subclasses_of` would be clearer and match the description of the method?

**Review by egiurleo — APPROVED** [review]
A name suggestion but otherwise this is super cool!

File: lib/spoom/deadcode/plugins/base.rb (hunk 1 above)
Comment (Morriar) [reply lib/spoom/deadcode/plugins/base.rb]: @st0012 raised the same concern about the name and we went with `ignore_classes_inheriting_from`.

But I agree with you Emily, the "direct" seems like an important information here.

On the other hand, maybe I can remove this restriction soon by computing the full type hierarchy 🤔

Comment (Morriar) [reply lib/spoom/deadcode/plugins/base.rb]: Maybe I should also change the name of `ignnore_class_names(...)` to `ignore_classes_named(...)` or something similar?

Comment (Morriar) [reply lib/spoom/deadcode/plugins/base.rb]: https://github.com/Shopify/spoom/pull/442

File: lib/spoom/deadcode/plugins/graphql.rb [hunk 2]
```diff
@@ -5,6 +5,13 @@ module Spoom
   module Deadcode
     module Plugins
       class GraphQL < Base
+        ignore_classes_inheriting_from(
+          "GraphQL::Schema::Enum",
```
[the remark this answers is not in the document]
Comment (Morriar) [reply lib/spoom/deadcode/plugins/graphql.rb:9]: Good call, they should 👍

[the remark this answers is not in the document]
Comment (st0012) [reply lib/spoom/deadcode/plugins/graphql.rb]: If all classes need to consider `::`, and the only use case of regexp args is for it, should the DSL automatically handle it in string form? Like `ignore_classes_inheriting_from("Foo")` handles both `Foo` and `::Foo`.
Have you experienced cases that only need one of them?

[the remark this answers is not in the document]
Comment (Morriar) [reply lib/spoom/deadcode/plugins/graphql.rb]: I think you right, in the string version we should normalize everything so the developer doesn't have to think about it, I'll push a change 👍

[the remark this answers is not in the document]
Comment (Morriar) [reply lib/spoom/deadcode/plugins/graphql.rb]: @st0012: done in https://github.com/Shopify/spoom/pull/445 👍

**Pull request merged**